WIP: OpenShift auth handler for V2 registry

pravisankar opened the pull request.

stevvooe committed:
spec: Canonicalized JSON Format for Distribution APIs

Signed-off-by: Stephen J Day <stephen.day@docker.com>

Commit:
Refactoring cloudfactory layer handler into a more generic storage
middleware concept.

This also breaks the dependency the storage package had on goamz
Signed-off-by: David Lawrence <david.lawrence@docker.com> (github: endophage)

Commit:
OpenShift auth handler for V2 registry

Commit:
Middleware!

Convert middleware in the config to be a map of type->[]Middleware

Add support for registry & repository middleware.

Some naming updates as well.

Signed-off-by: Andy Goldstein <agoldste@redhat.com>

Commit:
Final polish to cloudfront and larger middleware refactor
Signed-off-by: David Lawrence <david.lawrence@docker.com> (github: endophage)

Commit:
Fix Godoc typos

Signed-off-by: Andy Goldstein <agoldste@redhat.com>

Commit:
don't panic during a request when configuring repository middleware. Return a 500 with an appropriate error
Signed-off-by: David Lawrence <david.lawrence@docker.com> (github: endophage)

Commit:
supplement of digest refactoring

stevvooe committed:
Merge pull request #255 from xiekeyang/digest-refactoring

supplement of digest refactoring

ncdc commented on cmd/registry/main.go:
```diff
@@ -22,6 +22,7 @@ import (
 	gorhandlers "github.com/gorilla/handlers"
 	"github.com/yvasiyarov/gorelic"
 	"golang.org/x/net/context"
+	_ "../../registry/auth/openshift"
```
No need to fix this, but for future reference it's better to avoid relative paths in imports

ncdc commented on registry/auth/openshift/accessController.go:
```diff
@@ -0,0 +1,148 @@
+
+type openshiftToken struct {
+	namespace   string
+	imageName   string
```
this is imageRepo, not imageName

ncdc commented on registry/auth/openshift/accessController.go:
```diff
@@ -0,0 +1,148 @@
+var _ auth.AccessController = &accessController{}
+
+func (ac *accessController) verifyOpenShiftToken(token openshiftToken) error {
+	url := fmt.Sprintf("https://%s/subjectAccessReviews?namespace=%s", ac.openshiftAddr, token.namespace)
```
We should probably switch to using the Client from origin and use SubjectAccessReviewNamespacer, etc. But that can wait.

ncdc commented on registry/auth/openshift/accessController.go:
```diff
@@ -0,0 +1,148 @@
+			return nil, fmt.Errorf("Repository namespace not found for resource: %s", access.Resource.Name)
+		}
+		var token openshiftToken
+		token.namespace, token.imageName = repoParts[0], repoParts[1]
```
token := openshiftToken{namespace: repoParts[0], imageRepo: repoParts[1]}

ncdc commented on registry/auth/openshift/accessController.go:
```diff
@@ -0,0 +1,148 @@
+		var token openshiftToken
+		token.namespace, token.imageName = repoParts[0], repoParts[1]
+
+		if access.Action == "push" {
```
We tend to be using switch instead of if/else if/else

ncdc commented on registry/auth/openshift/accessController.go:
```diff
@@ -0,0 +1,148 @@
+			return nil, fmt.Errorf(`"Authorization" header not set`)
+		}
+		authParts := strings.Split(authorization, " ")
+		if len(authParts) != 2 || strings.ToLower(authParts[0]) != "bearer" {
```
We're not going to be able to support "Authorization: bearer" in the short term, because the Docker daemon implements a very specific workflow for bearer tokens (https://github.com/docker/distribution/blob/master/doc/spec/auth/token.md).

The flow will need to be:
1. acquire oauth token from OpenShift
2. `docker login`, and when it asks for password, specify the oauth token received in step 1
3. when you `docker pull` or `docker push`, docker will send the username and "password" (i.e. token) in an `Authorization: basic` header

ncdc commented on registry/auth/openshift/accessController.go:
```diff
@@ -0,0 +1,148 @@
+
+		authorization := req.Header.Get("Authorization")
+		if len(authorization) == 0 {
+			return nil, fmt.Errorf(`"Authorization" header not set`)
```
Instead of doing this, you'll need to create a challenge to send back to the client (Docker) so it knows it needs to authenticate. See https://github.com/docker/distribution/blob/master/registry/auth/silly/access.go#L52-L66 for an example.

An example challenge implementation is here: https://github.com/docker/distribution/blob/master/registry/auth/silly/access.go#L72-L91. Note, however, you'll need to send back a "basic" challenge and not a "bearer" one.

Commit:
Send WWW-Authenticate header for silly auth

Signed-off-by: Andy Goldstein <agoldste@redhat.com>

stevvooe committed:
Merge pull request #244 from endophage/cloudfront_refactor

registry/middleware, registry/storage, configuration: refactoring cloudfront + generic middlewares

stevvooe committed:
Merge pull request #256 from ncdc/master

Send WWW-Authenticate header for silly auth

Commit:
Insert request method option storage driver URLFor

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

stevvooe committed:
Merge pull request #258 from jlhawn/url_for_method

Insert request method option storage driver URLFor

stevvooe committed:
Validate prefix lookups for WithRequest/WithResponse contexts

Signed-off-by: Stephen J Day <stephen.day@docker.com>

Commit:
Refactor Layer interface to return a Handler

... Rather than ServeHTTP directly.

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

dmp42 committed:
Merge pull request #259 from stevvooe/validate-var-prefixes

context: validate prefix lookups for WithRequest/WithResponse contexts

stevvooe committed:
Merge pull request #261 from jlhawn/url_for_method

Refactor Layer interface to return a Handler

stevvooe committed:
Merge pull request #226 from stevvooe/json-canonicalization

doc/spec: canonicalized JSON format for distribution APIs
